File: backend/services/prediction_service.py
"""
=========================================================
Prediction Service
=========================================================
"""

from sqlalchemy.orm import Session

from database.crud.prediction_crud import (
    create_prediction,
    get_prediction,
    get_predictions,
    update_prediction,
    delete_prediction,
)

from database.schemas.prediction_schema import (
    PredictionCreate,
    PredictionUpdate,
)

from backend.utils.exceptions import PredictionNotFound


class PredictionService:

    @staticmethod
    def create(
        db: Session,
        prediction: PredictionCreate,
    ):
        return create_prediction(db, prediction)

    @staticmethod
    def get(
        db: Session,
        prediction_id: int,
    ):
        prediction = get_prediction(db, prediction_id)

        if prediction is None:
            raise PredictionNotFound()

        return prediction

    @staticmethod
    def get_all(
        db: Session,
    ):
        return get_predictions(db)

    @staticmethod
    def update(
        db: Session,
        prediction_id: int,
        prediction_update: PredictionUpdate,
    ):
        prediction = update_prediction(
            db,
            prediction_id,
            prediction_update,
        )

        if prediction is None:
            raise PredictionNotFound()

        return prediction

    @staticmethod
    def delete(
        db: Session,
        prediction_id: int,
    ):
        prediction = delete_prediction(
            db,
            prediction_id,
        )

        if prediction is None:
            raise PredictionNotFound()

        return prediction

File: frontend/src/services/predictionService.js
import api from "../api/api";

// ML Prediction
export const predict = async (data) => {
  const response = await api.post("/ml/predict", data);
  return response.data;
};

// Get all saved predictions
export const getPredictions = async () => {
  const response = await api.get("/predictions");
  return response.data;
};

// Get prediction by ID
export const getPrediction = async (id) => {
  const response = await api.get(`/predictions/${id}`);
  return response.data;
};

// Update prediction
export const updatePrediction = async (id, data) => {
  const response = await api.put(`/predictions/${id}`, data);
  return response.data;
};

// Delete prediction
export const deletePrediction = async (id) => {
  const response = await api.delete(`/predictions/${id}`);
  return response.data;
};